src/fingerprint.cpp

```diff
@@ -114,6 +135,84 @@ std::string compute_uri_extention(const std::string& path) {
     return ext;
 }
 
+void decode(const std::string& str, std::string& decodedStr) {
+    std::string::const_iterator it = str.begin();
+    std::string::const_iterator end = str.end();
+    const int OFFSET = 10;
+    const int HEX_OFFSET = 16;
+    while (it != end) {
+        int c = *it++;
+        if (c == '%') {
+            if (it == end) {
+                throw std::exception();
+            }
+            int hi = *it++;
+            if (it == end) {
+                throw std::exception();
+            }
+            int lo = *it++;
+            if (hi >= '0' && hi <= '9') {
+                c = hi - '0';
+            } else if (hi >= 'A' && hi <= 'F') {
+                c = hi - 'A' + OFFSET;
+            } else if (hi >= 'a' && hi <= 'f') {
+                c = hi - 'a' + OFFSET;
+            } else {
+                throw std::exception();
+            }
+            c *= HEX_OFFSET;
+            if (lo >= '0' && lo <= '9') {
+                c += lo - '0';
+            } else if (lo >= 'A' && lo <= 'F') {
+                c += lo - 'A' + OFFSET;
+            } else if (lo >= 'a' && lo <= 'f') {
+                c += lo - 'a' + OFFSET;
+            } else {
+                throw std::exception();
+            }
+        }
+        decodedStr += std::to_string(c);
+    }
+}
+
+std::vector<std::pair<std::string, std::string>> get_query_parameters(const std::string& query) {
+    std::vector<std::pair<std::string, std::string>> result;
+    std::string::const_iterator it(query.begin());
+    std::string::const_iterator end(query.end());
+    while (it != end) {
+        std::string name;
+        std::string value;
+        while (it != end && *it != '=' && *it != '&') {
+            if (*it == '+') {
+                name += ' ';
+            } else {
+                name += *it;
+            }
+            ++it;
+        }
+        if (it != end && *it == '=') {
+            ++it;
+            while (it != end && *it != '&') {
+                if (*it == '+') {
+                    value += ' ';
+                } else {
+                    value += *it;
+                }
+                ++it;
+            }
+        }
+        std::string decodedName;
+        std::string decodedValue;
+        decode(name, decodedName);
+        decode(value, decodedValue);
+        result.push_back(std::make_pair(decodedName, decodedValue));
+        if (it != end && *it == '&') {
+            ++it;
+        }
+    }
+    return result;
+}
```

Comment:
It is not necessary but it could be nice to have a reference to the original source code from POCO as a comment